core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java

```diff
@@ -0,0 +1,131 @@
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+
+import java.util.Objects;
+
+import static org.elasticsearch.action.support.replication.ReplicationOperation.WAIT_ON_ACTIVE_SHARDS_SETTING;
+import static org.elasticsearch.action.support.replication.ReplicationOperation.checkWaitOnActiveShards;
+
+/**
+ * This class is responsible for waiting to send an index creation
+ * response on the action listener until indexing or other replication
+ * actions can successfully be performed on the newly created index.
+ *
+ * It should be subclassed to implement the particular response type
+ * desired for the index creation operation.
+ */
+public final class IndexCreationMonitor extends AbstractComponent {
+
+    private final ClusterService clusterService;
+    private final ThreadContext threadContext;
+
+    protected IndexCreationMonitor(Settings settings, ClusterService clusterService, ThreadContext threadContext) {
+        super(settings);
+        this.clusterService = Objects.requireNonNull(clusterService);
+        this.threadContext = Objects.requireNonNull(threadContext);
+    }
+
+    void waitOnShards(final String indexName,
+                      final int waitOnActiveShards,
+                      final ActionListener<ClusterStateUpdateResponse> listener,
+                      final TimeValue timeout) {
+
+        final ClusterStateObserver observer = new ClusterStateObserver(clusterService, logger, threadContext);
+        final ClusterState observedState = observer.observedState();
+        final ClusterStateObserver.ChangePredicate shardsAllocatedPredicate = new ClusterStateObserver.ChangePredicate() {
+            @Override
+            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                return newStatus == ClusterState.ClusterStateStatus.APPLIED
+                           && enoughShardsActive(newState, indexName, settings, waitOnActiveShards);
+            }
+
+            @Override
+            public boolean apply(ClusterChangedEvent changedEvent) {
+                return enoughShardsActive(changedEvent.state(), indexName, settings, waitOnActiveShards);
+            }
+        };
+        final ClusterStateObserver.Listener observerListener = new ClusterStateObserver.Listener() {
+            @Override
+            public void onNewClusterState(ClusterState state) {
+                listener.onResponse(new ClusterStateUpdateResponse(true));
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                logger.warn("[{}] cluster service closed while waiting for write consistency number of shards " +
+                                "to be allocated on index creation.", indexName);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
```

Comment:
We should notify the client about the node closing instead: `listener.onFailure(new NodeClosedException(clusterService.localNode()))`